lib/fs.js

```diff
@@ -243,20 +247,26 @@ fs.readFile = function(path, options, callback_) {
     return;
 
   var context = new ReadFileContext(callback, encoding);
+  context.isUserFd = isFd(path); // file descriptor ownership
   var req = new FSReqWrap();
   req.context = context;
   req.oncomplete = readFileAfterOpen;
 
-  binding.open(pathModule._makeLong(path),
-               stringToFlags(flag),
-               0o666,
-               req);
+  if (context.isUserFd) {
+    req.oncomplete.call(req, null, path);
```

Comment:
Put this in a `nextTick`. It's an expected async call, and must always run async.